### ISSUE TITLE
We need to fix slow CI ASAP

### ISSUE THREAD

**Issue by camsaul** [metabase/metabase#12437]
CI took [about 9 minutes to run](https://app.circleci.com/pipelines/github/metabase/metabase/4936/workflows/312e5170-0eff-4968-b3ed-a55bb770524b) only a few months ago (which is already a bit long). Today, [it takes 30 minutes](https://app.circleci.com/pipelines/github/metabase/metabase/7445/workflows/796a0fe3-4f93-4971-bda5-aa8af89bba1c) for a simple run; runs with `[ci all]`/`[ci drivers]` [take around ~45 minutes](https://app.circleci.com/pipelines/github/metabase/metabase/7402/workflows/daeacf40-afbe-4295-8e16-487e7b89570c). 

Over the past few years I've spent a lot of time tuning CI and trying to keep run times short with all sorts of smart caching and things like that. We need to revisit stuff that was added recently and apply some of the same optimizations.

The culprits are:

*  The Cypress tests. This step took 11 minutes on the test run linked above. It seems to be going up by a minute or so each month, so if we don't do something about it CI is going to take 40 minutes before the year is out
*  Changes in #12292 that now build the all drivers before building the uberjar to allow FE tests for MongoDB. Previously, the `build-uberjar` step took 2 minutes; it now takes 18

I'm not proposing we remove the new test suites, but we **need** to optimize them and make CI usable again. Cypress tests can't run until the uberjar is built, so building the uberjar + running cypress now takes 29 minutes alone.

Here's some fix ideas:
*  Cache the built drivers so we don't need to rebuild them unless their source changes
*  Split cypress tests up and run a few different suites in parallel
*  Improve logic for automatically skipping steps if relevant files are unchanged 
*  Make uberjar w/ drivers and/or cypress tests `[ci all]` if we can't make them fast. That's what we do with driver tests

On an unrelated note I've noticed [the `fe-deps` step is taking close to a full minute now](https://app.circleci.com/pipelines/github/metabase/metabase/7445/workflows/796a0fe3-4f93-4971-bda5-aa8af89bba1c/jobs/246739), mostly because the cache getting restored is enormous -- this will continue to get worse because the cache keeps accumulating packages and never shrinks in size. This same cache gets restored before running all FE steps, so it adds as much as a minute to each step. A couple ideas to fix this:
*  We could tweak the cache key so it includes current month or something. That will effectively force a cache flush every month
*  If there's some way to remove unused packages `.yarn`/`.yarn_cache`/`node_modules`/etc. after fetching deps we should do that, that way unused stuff isn't saved to the cache

**Comment by dacort** [comment]
Is there a way to break up the tests such that changes are isolated?

**Comment by camsaul** [comment]
we already have logic in place that calculates a checksum of backend files and uses that as a cache key for the uberjar. That way if the backend files are unchanges it can use the cached uberjar instead of rebuilding it.

The script that builds drivers actually does something similar. It stores artifacts in `resources/modules` and stores the checksums in `modules/drivers/<driver>/target/checksum.txt`. So if we were to cache those files then we could avoid rebuilding the drivers pretty easily.

Basically we'd want to restore the cached files at the beginning of the `build-uberjar` step and then save those files again at the end of the `build-uberjar` step

**Comment by camsaul** [comment]
You can't disable a build step completely in Circle but you can have the step exit with 0 before it does anything important, which is how we implement the driver tests only running if you include `[ci drivers]`/`[ci all]`/`[ci <driver>]`/run on `master` or `release-*`.

So we *could* do something similar for the FE tests

**Comment by camsaul** [comment]
A slightly more advanced option would be to do a checksum of the relevant files and then cache a dummy file for a step if the step completes using that checksum as a key.

e.g. and example for a backend test step could look like:

1.  Calculate `checksum` of all backend files and restore files from the cache with key `my-backend-step-{{checksum}}` if it exists
2.  If a file named `SUCCESS` if present, `exit 0`, if not, `lein test`
3.  `touch SUCCESS`
4.  cache `SUCCESS` with the key `my-backend-step-{{checksum}}` 

So that way, if we're running the step against a given checksum, it will effectively no-op if it passed with the same files already.

Unfortunately for Cypress I think basically any backend or frontend file would have to be included in the checksum since they're all potentially relevant

**Comment by camsaul** [comment]
I think if we
*  prune the FE deps cache
*  make sure the build-uberjar step is caching drivers instead of rebuilding them every time the uberjar is built
*  shard the cypress tests 

We can get normal CI runs back under the sub-10-minute mark.

**Comment by dacort** [comment]
Sounds pretty reasonable.

We do need to be careful on the last point as we may have some shared state in a few of our Cypress tests, which is not [best practice](https://docs.cypress.io/guides/references/best-practices.html#Having-tests-rely-on-the-state-of-previous-tests).

**Comment by tlrobinson** [comment]
Just to be explicit:

1. we can't have multiple tests running in parallel against the same backend because they reset the app db before each test, but we should be able to run multiple Circle Nodes
2. for the reason @dacort mentioned, the sharding must occur at the spec file level rather than the individual test level

I think it should be fine on both counts. We need to set up the Cypress dashboard and recording

> Cypress will automatically balance your spec files across the available machines in your CI provider. Cypress calculates which spec file to run based on the data collected from previous runs. This ensures that your spec files run as fast as possible, with no need for manual configuration.
>
> As more and more tests are recorded to the Cypress Dashboard, Cypress can better predict how long a given spec file will take to run. To prevent irrelevant data from affecting the duration prediction, Cypress doesn’t use old historical run data regarding the spec file.

**Comment by camsaul** [comment]
@tlrobinson that's what I had in mind, splitting the tests up into 2 or 4 circle tasks. So all completely separate

**Comment by tlrobinson** [comment]
Also I'm not sure how much time it would save, but we could run Cypress tests against a backend started with `lein run` (though we'd still need to build the FE, and drivers for cypress-mongo)

**Comment by camsaul** [comment]
@tlrobinson since we wouldn't have to build the uberjar it seems like it would save 20+ minutes

**Comment by camsaul** [comment]
Fixed by #12620, #13924, #13942
